src/main/scala/com/fulcrumgenomics/bam/ConsensusCaller.scala

```diff
@@ -0,0 +1,369 @@
+    val pr1 = prErrorTrialOne       * prErrorTrialTwo.inv()
+    val pr2 = prErrorTrialOne.inv() * prErrorTrialTwo
+    val pr3 = prErrorTrialOne       * prErrorTrialTwo       * 2.0 / 3.0
+    pr1 + pr2 + pr3
+  }
+
+  /** The type of consensus read to output. */
+  private object ReadType extends Enumeration {
+    val Fragment, FirstOfPair, SecondOfPair = Value
+  }
+
+  /** Gets the longest common prefix of the given strings, None if there is only one string or if there is an empty string. */
+  @annotation.tailrec
+  private def longestCommonPrefix(strs: Iterable[String], accu: Option[String] = None): Option[String] = {
+    if (strs.exists(_.isEmpty) || strs.size <= 1) accu
+    else {
+      val firsts = strs.map(_.head)
+      if (firsts.toSet.size > 1) accu
+      else longestCommonPrefix(strs.map(_.tail), Some(accu.getOrElse("") + firsts.head))
+    }
+  }
+}
+
+/** Calls consensus reads by grouping consecutive reads with the same attribute.
+  *
+  * Consecutive reads with the sam attribute are partitioned into fragments, first of pair, and
+  * second of pair reads, and a consensus read is created for each partition.  A consensus read
+  * for a given partition may not be returned if any of the conditions are not met (ex. minimum
+  * number of reads, minimum mean consensus base quality, ...).
+  * */
+class ConsensusCaller
+( input: Iterator[SAMRecord],
+  val header: SAMFileHeader,
+  val readNamePrefix: Option[String]   = None,
+  val readGroup: String                = "A",
+  val options: ConsensusCallerOptions  = new ConsensusCallerOptions(),
+  val rejects: Option[SAMFileWriter]   = None,
+  val progress: Option[ProgressLogger] = None
+) extends Iterator[SAMRecord] {
+  import ConsensusCaller.ReadType
+  import ConsensusCaller.ReadType._
+  import ConsensusCaller._
+
+  private val iter = input.buffered
+  private val nextConsensusRecords: mutable.Queue[SAMRecord] = mutable.Queue[SAMRecord]() // one per UMI group
+  private var readIdx = 1
+
+  /** True if there are more consensus reads, false otherwise. */
+  def hasNext(): Boolean = {
+    if (this.nextConsensusRecords.nonEmpty) true
+    else this.advance()
+  }
+
+  /** Returns the next consensus read. */
+  def next(): SAMRecord = {
+    if (!this.hasNext()) throw new IllegalStateException("Calling next() when hasNext() is false.")
+    this.nextConsensusRecords match {
+      case mutable.Queue(x, _*) => this.nextConsensusRecords.dequeue()
+      case mutable.Queue() => unreachable("hasNext() was true but nextConsensusRecords is empty")
+    }
+  }
+
+  /** Consumes records until a consensus read can be created, or no more input records are available. Returns
+    * true if a consensus read was created, false otherwise. */
+  private def advance(): Boolean = {
+    if (this.iter.isEmpty) return false
+
+    // get the records to create the consensus read
+    val buffer = getNextGroupOfRecords(attr=this.iter.head.toAttr)
```

Comment:
I see, with your `getNextGroupOfRecords` re-implementation below, the point is made.